std/experimental/logger/asynclogger.d

```diff
@@ -0,0 +1,127 @@
+module std.experimental.logger.asynclogger;
+
+import std.experimental.logger.core;
+import std.concurrency;
+
+/** The $(D AsyncLogger) processes logs asynchronously by using another $(D Logger) implementation
+as a backend. Every data logged to this $(D AsyncLogger) will be send to separate logging thread
+in order to be logged there. It can help to improve logging performance for $(D log) caller,
+for example by removing waiting for flushing logs to a file.
+Example:
+-------------
+void main()
+{
+    import std.algorithm;
+    import std.conv;
+    import std.stdio;
+    import std.datetime;
+    import std.experimental.logger;
+
+    auto fl = new FileLogger("defaultFileLogger.log", LogLevel.all);
+    void logFileDefault()
+    {
+        fl.log("test message");
+    }
+
+    auto afl = new AsyncLogger(new FileLogger("asyncFileLogger.log", LogLevel.all));
+    void logFileAsync()
+    {
+        afl.log("test message");
+    }
+
+    auto r = benchmark!(logFileDefault, logFileAsync)(100_000);
+    writeln(map!(a => to!string(to!Duration(a)))(r[]));
+}
+-------------
+*/
+final class AsyncLogger : Logger
+{
+    /// Shuntdown message
+    struct ShutdownMsg {}
+
+    private static void spawnedFunc(shared Logger ptr) @trusted
+    {
+        auto l = cast(Logger)(ptr);
+        bool isRunning = true;
+        while (isRunning)
+        {
+            receive(
+                (LogEntryBase e)
+                {
+                    LogEntry tmp = e;
+                    tmp.threadId = ownerTid;
+                    tmp.logger = l;
+                    l.forwardMsg(tmp);
+                },
+                (ShutdownMsg msg) { isRunning = false; },
+                (OwnerTerminated ex) { isRunning = false; },
+                (Variant v) { throw new Error("unhandled message"); }
+            );
+        }
+    }
+
+    /** This constructor sets a LogLevel and starts a new logging thread
+    that logs every entry asynchronously by using Logger. Because logger writes down
+    all entries into special logging thread, logger should not be used further
+    (without confidence that logger is not being used at the moment).
+    */
+    this(Logger logger, LogLevel lv = LogLevel.all) @trusted
+    {
+        super(lv);
+        this.logger_ = cast(shared Logger)(logger);
+        this.tid_ = spawn(&spawnedFunc, this.logger_);
+    }
+
+    override void writeLogMsg(ref LogEntry payload) @trusted
+    {
+        LogEntryBase msg = payload;
+        send(this.tid_, msg);
+    }
+
+    /// Stops logging thread by sending ShutdownMsg to it.
+    ~this() @trusted
+    {
+        send(this.tid_, ShutdownMsg());
+    }
+
+    Tid tid_;
```

Comment:
> There should be a comment on how to use this Tid to stop the thread and/or restart it

Horrible idea, an active object [[¹]](www.drdobbs.com/parallel/prefer-using-active-objects-instead-of-n/225700095) doesn't leak it's threading details. I'm not sure what you mean with restart. The only question is how to deal with exceptions of the underlying logger. You could rethrow them on the next call to AsyncLogger.